src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlQueueClient.cs

```diff
@@ -152,137 +150,101 @@ public async Task<IReadOnlyCollection<JobInfo>> DequeueJobsAsync(byte queueType,
-                if (groupId.HasValue)
+            return await _sqlRetryService.ExecuteSqlCommandFuncWithRetries(
+                async (sqlCommand, cancellationToken) =>
                 {
-                    sqlCommandWrapper.Parameters.AddWithValue("@GroupId", groupId.Value);
-                }
-
-                sqlCommandWrapper.Parameters.AddWithValue("@ForceOneActiveJobGroup", forceOneActiveJobGroup);
-                sqlCommandWrapper.Parameters.AddWithValue("@IsCompleted", isCompleted);
+                    // cannot use VLatest as it does not understand optional parameters
+                    sqlCommand.CommandType = System.Data.CommandType.StoredProcedure;
+                    sqlCommand.CommandText = "dbo.EnqueueJobs";
+                    sqlCommand.Parameters.AddWithValue("@QueueType", queueType);
+                    new StringListTableValuedParameterDefinition("@Definitions").AddParameter(sqlCommand.Parameters, definitions.Select(d => new StringListRow(d)));
+                    if (groupId.HasValue)
+                    {
+                        sqlCommand.Parameters.AddWithValue("@GroupId", groupId.Value);
+                    }
 
-                try
-                {
-                    await using SqlDataReader sqlDataReader = await sqlCommandWrapper.ExecuteReaderAsync(cancellationToken);
+                    sqlCommand.Parameters.AddWithValue("@ForceOneActiveJobGroup", forceOneActiveJobGroup);
+                    sqlCommand.Parameters.AddWithValue("@IsCompleted", isCompleted);
 
-                    return await sqlDataReader.ReadJobInfosAsync(cancellationToken);
-                }
-                catch (SqlException sqlEx)
-                {
-                    if (sqlEx.State == 127)
+                    try
                     {
-                        throw new JobManagement.JobConflictException(sqlEx.Message);
-                    }
+                        await using SqlDataReader sqlDataReader = await sqlCommand.ExecuteReaderAsync(cancellationToken);
 
-                    throw;
-                }
-            }
-            catch (Exception ex)
-            {
-                _logger.LogError(ex, "EnqueueAsync failed.");
-                if (ex.IsRetryable())
-                {
-                    throw new RetriableJobException(ex.Message, ex);
-                }
+                        return await sqlDataReader.ReadJobInfosAsync(cancellationToken);
+                    }
+                    catch (SqlException sqlEx)
+                    {
+                        if (sqlEx.State == 127)
+                        {
+                            throw new JobManagement.JobConflictException(sqlEx.Message);
+                        }
 
-                throw;
-            }
+                        throw;
+                    }
+                },
+                _logger,
+                "EnqueueAsync failed.",
```

Comment:
>"EnqueueAsync failed.", [](http://example.com/codeflow?start=16&length=23)

This message can be predicted. Why do we need this parameter?

Reply:
Logging message if retries. Info log for each retry except last one which is log error.